include/sys/async_notify.h

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ASYNC_NOTIFY_H_
+#define ZEPHYR_INCLUDE_SYS_ASYNC_NOTIFY_H_
+
+#include <kernel.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup async_notify_apis Asynchronous Notification APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/* Forward declaration */
+struct async_notify;
+
+/*
+ * Flag value that overwrites the method field when the operation has
+ * completed.
+ */
+#define ASYNC_NOTIFY_METHOD_COMPLETED 0
```

Comment:
Again, I disagree with these long names. I really would like to shorten this to something like `ANOTIFY_` and `anotify_` or something even shorter, if we can come up with one. I will add this to the API meeting.

Reply:
The whole namespace would need to change. I don't know if we need to keep async in the name, notify_* only perhaps?

Reply:
in a way, by default notification is asynchronous in nature so `async_` prefix might be redundant. 
Some other proposals :
- `notifier`
- `gnotify` - generic notification
- `k_notify` - since (a least currently) it is limited to kernel.

Reply:
gnotify sounds like glib, your are right k_ is strictly used by the kernel only. z_notify maybe? Though it does not "sound" nice.

Reply:
yeah, i thought about z_notify but `z_` indicates internal function.

Reply:
I suppose that `async_` doesn't add a huge amount of information except that it's intended to be the standard way of notifying of completion of async operations.

Bare `notify_` IMO has too large of a conflict with existing and new identifiers that are not part of this API (`struct notify_data`, `notify_remote_info`).

Since this is in the `sys/` include hierarchy `sys_notify_` could mitigate that.

Reply:
API meeting:
`sys_notify_` raised no objections.

Reply:
how to we call the file? `notify.h`? since path contains the namespace (`sys/notify.h`)?

Reply:
yes, as all the other files there. (besides sys_io.h which is a legacy name, and was not part of sys/ at first if I remember well)